File: src/MultiAgentApp/MCP/McpClientFactory.cs
using Microsoft.Extensions.AI;
using Microsoft.Extensions.Logging;
using ModelContextProtocol.Client;
using MultiAgentApp.Configuration;

namespace MultiAgentApp.MCP;

/// <summary>
/// Creates MCP (Model Context Protocol) clients and converts their tools to
/// <see cref="AITool"/> instances that Microsoft Agent Framework agents can call.
///
/// Transport strategy:
/// • Local (default): launches server executables as child processes via stdio transport.
/// • Azure APIM: switches to an HTTP SSE transport once Azure API Management MCP endpoints
///   become available. Toggle <c>MCP:UseAzureApim</c> in configuration.
/// </summary>
public sealed class McpClientFactory : IAsyncDisposable
{
    private readonly McpOptions _options;
    private readonly ILogger<McpClientFactory> _logger;
    private readonly List<McpClient> _clients = [];

    public McpClientFactory(McpOptions options, ILogger<McpClientFactory> logger)
    {
        _options = options;
        _logger = logger;
    }

    /// <summary>
    /// Connects to the Weather MCP server and returns its tools as <see cref="AITool"/> objects.
    /// </summary>
    public async Task<IList<AITool>> GetWeatherToolsAsync(CancellationToken ct = default)
        => await GetToolsAsync(_options.WeatherServer, ct);

    /// <summary>
    /// Connects to the Products MCP server and returns its tools as <see cref="AITool"/> objects.
    /// </summary>
    public async Task<IList<AITool>> GetProductsToolsAsync(CancellationToken ct = default)
        => await GetToolsAsync(_options.ProductsServer, ct);

    // ── Private helpers ──────────────────────────────────────────────────────

    private async Task<IList<AITool>> GetToolsAsync(McpServerOptions serverOptions, CancellationToken ct)
    {
        IClientTransport transport = _options.UseAzureApim
            ? BuildApimTransport(serverOptions)
            : BuildStdioTransport(serverOptions);

        _logger.LogInformation(
            "Connecting to MCP server '{Name}' via {Mode} transport.",
            serverOptions.Name,
            _options.UseAzureApim ? "APIM" : "stdio");

        var client = await McpClient.CreateAsync(transport, cancellationToken: ct);
        _clients.Add(client);

        var tools = await client.ListToolsAsync(cancellationToken: ct);

        // McpClientTool inherits from AIFunction which inherits from AITool,
        // so it can be used directly as AITool by MAF agents.
        return [.. tools];
    }

    /// <summary>
    /// Creates a stdio transport that launches the server as a child process.
    /// Used during local development and CI.
    /// </summary>
    private static IClientTransport BuildStdioTransport(McpServerOptions opts) =>
        new StdioClientTransport(new StdioClientTransportOptions
        {
            Name = opts.Name,
            Command = opts.StdioCommand,
            Arguments = opts.StdioArguments
        });

    /// <summary>
    /// Creates an HTTP SSE transport that connects to Azure API Management.
    /// Replace this stub with the actual APIM SSE URL once available.
    /// </summary>
    private static IClientTransport BuildApimTransport(McpServerOptions opts)
    {
        // ── Azure APIM MCP Transport ────────────────────────────────────────────
        // Uncomment once Azure API Management MCP support is available:
        //
        // return new SseClientTransport(new SseClientTransportOptions
        // {
        //     Endpoint = new Uri(opts.ApimEndpoint),
        //     AdditionalHeaders = new Dictionary<string, string>
        //     {
        //         ["Ocp-Apim-Subscription-Key"] = opts.ApimApiKey
        //     }
        // });
        // ── End Azure APIM MCP Transport ────────────────────────────────────────

        throw new NotImplementedException(
            $"Azure APIM transport is not yet implemented for '{opts.Name}'. " +
            "Set MCP:UseAzureApim=false to continue using local stdio transport.");
    }

    public async ValueTask DisposeAsync()
    {
        foreach (var client in _clients)
        {
            await client.DisposeAsync();
        }
        _clients.Clear();
    }
}

File: src/multi_agent_app/mcp/mcp_client_factory.py
"""MCP (Model Context Protocol) client factory.

Connects to MCP servers via stdio transport and exposes their tools as
OpenAI-compatible function definitions, together with an async executor that
calls the live MCP session.

Transport strategy
------------------
• Local (default): launches server executables as child processes via stdio.
• Azure APIM: switches to an HTTP/SSE transport once Azure API Management MCP
  endpoints become available.  Toggle ``MCP.UseAzureApim`` in settings.
"""

from __future__ import annotations

import logging
import os
from collections.abc import Awaitable, Callable
from contextlib import AsyncExitStack
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

from multi_agent_app.config.mcp_options import McpOptions, McpServerOptions

logger = logging.getLogger(__name__)

# Add the ``src/`` directory to PYTHONPATH of spawned MCP-server subprocesses
# so they can import their own packages as top-level modules.
_SRC_DIR = str(Path(__file__).parent.parent.parent)


@dataclass
class McpToolSet:
    """Tool definitions and their live MCP executor for one server.

    Attributes:
        definitions: List of tool schemas in OpenAI function-call format.
        executor: Async callable ``(tool_name, args_dict) -> result_str`` that
            calls the live MCP session.  ``None`` means no tools are available.
    """

    definitions: list[dict[str, Any]] = field(default_factory=list)
    executor: Callable[[str, dict[str, Any]], Awaitable[str]] | None = None


class McpClientFactory:
    """Creates and manages MCP client connections.

    Use as an async context manager to ensure all stdio sessions are cleanly
    torn down when the application exits::

        async with McpClientFactory(mcp_options) as factory:
            weather_tools = await factory.get_weather_tools()
            ...

    The ``McpToolSet.executor`` closures remain valid for the lifetime of the
    ``async with`` block.
    """

    def __init__(self, options: McpOptions) -> None:
        self._options = options
        self._exit_stack = AsyncExitStack()

    async def __aenter__(self) -> McpClientFactory:
        await self._exit_stack.__aenter__()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self._exit_stack.__aexit__(*args)

    async def get_weather_tools(self) -> McpToolSet:
        """Connect to the Weather MCP server and return its tools."""
        return await self._get_tools(self._options.weather_server)

    async def get_products_tools(self) -> McpToolSet:
        """Connect to the Products MCP server and return its tools."""
        return await self._get_tools(self._options.products_server)

    async def _get_tools(self, server_options: McpServerOptions) -> McpToolSet:
        if self._options.use_azure_apim:
            raise NotImplementedError(
                f"Azure APIM transport is not yet implemented for "
                f"'{server_options.name}'. "
                "Set MCP.UseAzureApim=false to continue using local stdio transport."
            )

        env = os.environ.copy()
        existing_pythonpath = env.get("PYTHONPATH", "")
        env["PYTHONPATH"] = (
            _SRC_DIR + os.pathsep + existing_pythonpath
            if existing_pythonpath
            else _SRC_DIR
        )

        server_params = StdioServerParameters(
            command=server_options.stdio_command,
            args=server_options.stdio_arguments,
            env=env,
        )

        logger.info(
            "Connecting to MCP server '%s' via stdio transport.",
            server_options.name,
        )

        read, write = await self._exit_stack.enter_async_context(
            stdio_client(server_params)
        )
        session: ClientSession = await self._exit_stack.enter_async_context(
            ClientSession(read, write)
        )
        await session.initialize()

        tools_result = await session.list_tools()
        tools = tools_result.tools

        definitions: list[dict[str, Any]] = [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description or "",
                    "parameters": t.inputSchema,
                },
            }
            for t in tools
        ]

        async def executor(name: str, args: dict[str, Any]) -> str:
            call_result = await session.call_tool(name, args)
            parts: list[str] = []
            for item in call_result.content:
                if hasattr(item, "text"):
                    parts.append(item.text)
                else:
                    parts.append(str(item))
            return "\n".join(parts)

        return McpToolSet(definitions=definitions, executor=executor)
